File: control-plane/src/server/mcp-tools/composeException.ts
import { defineTool } from "@github/copilot-sdk";
import { z } from "zod";
import { nanoid } from "nanoid";
import type { StateStore } from "@server/services/stateStore";
import type { EventBus } from "@server/services/eventBus";
import type { AuditLogger } from "@server/services/auditLogger";
import type { Exception, ExceptionCategory, ExceptionOption, PolicyRef } from "@shared/types";

export function composeExceptionTool(store: StateStore, _bus: EventBus, audit: AuditLogger) {
  return defineTool("compose-exception", {
    description: "Write an exception to the queue. Non-revocable action — audited before and after write.",
    parameters: z.object({
      workflowId: z.string().describe("The workflow this exception applies to"),
      severity: z.enum(["critical", "high", "medium"]).describe("Exception severity"),
      category: z.string().describe("Exception category (e.g. po-mismatch, duplicate-invoice)"),
      summary: z.string().describe("Concise summary of the exception for the operator"),
      recommendation: z.string().describe("Agent recommendation for resolving this exception"),
      options: z
        .array(
          z.object({
            label: z.string(),
            action: z.string(),
            nonRevocable: z.boolean(),
          })
        )
        .optional()
        .describe("Decision options to present to the operator (defaults to approve/reject)"),
      relatedPolicyRefs: z
        .array(
          z.object({
            title: z.string(),
            snippet: z.string(),
            source: z.string(),
          })
        )
        .optional()
        .describe("Policy references that informed this exception"),
      bulkCandidateIds: z
        .array(z.string())
        .optional()
        .describe("Other workflow IDs that are candidates for bulk resolution"),
      confidence: z
        .number()
        .min(0)
        .max(1)
        .optional()
        .describe("Agent confidence score between 0 and 1"),
    }),
    skipPermission: true,
    handler: async ({
      workflowId,
      severity,
      category,
      summary,
      recommendation,
      options,
      relatedPolicyRefs,
      bulkCandidateIds,
      confidence,
    }) => {
      // Hook-gated non-revocable action: audit BEFORE writing
      audit.log({
        action: "compose-exception.pre",
        details: { workflowId, severity, category, summary },
        timestamp: Date.now(),
      });

      const defaultOptions: ExceptionOption[] = [
        { label: "Approve", action: "approve", nonRevocable: true },
        { label: "Reject", action: "reject", nonRevocable: false },
      ];

      const exc: Exception = {
        id: `EXC-${nanoid(8)}`,
        workflowId,
        composedBy: "fleet-manager",
        severity,
        category: category as ExceptionCategory,
        summary,
        recommendation,
        options: options ?? defaultOptions,
        relatedPolicyRefs: (relatedPolicyRefs ?? []) as PolicyRef[],
        bulkCandidateIds,
        confidence: confidence ?? 0.8,
        createdAt: Date.now(),
      };

      store.upsertException(exc);

      audit.log({
        action: "compose-exception.emitted",
        details: { exceptionId: exc.id, workflowId, severity, category },
        timestamp: Date.now(),
      });

      return { exceptionId: exc.id };
    },
  });
}

File: control-plane-py/src/server/mcp_tools/compose_exception.py
# src/server/mcp_tools/compose_exception.py
from __future__ import annotations
import json
import time
from typing import Any
from pydantic import BaseModel, Field
from nanoid import generate as nanoid
from copilot.tools import define_tool, ToolInvocation, ToolResult
from src.server.services.state_store import StateStore
from src.server.services.audit_logger import AuditLogger
from src.shared.types import Exception_ as Exception, ExceptionOption, PolicyRef


class ComposeExceptionParams(BaseModel):
    workflow_id: str
    severity: str = Field(description="critical | high | medium")
    category: str = Field(description="Exception category enum value")
    summary: str
    recommendation: str
    options: list[dict[str, Any]] | None = None
    related_policy_refs: list[dict[str, Any]] | None = None
    bulk_candidate_ids: list[str] | None = None
    confidence: float = 0.8


def make_compose_exception_tool(store: StateStore, audit: AuditLogger):
    @define_tool(description="Write an exception to the operator's queue.", skip_permission=True)
    def compose_exception(params: ComposeExceptionParams, invocation: ToolInvocation) -> ToolResult:
        # Hook-gated non-revocable action: pre-audit BEFORE write.
        audit.log("compose-exception.pre", {"workflow_id": params.workflow_id})
        e = Exception(
            id=f"EXC-{nanoid(size=8)}",
            workflow_id=params.workflow_id,
            composed_by="fleet-manager",
            severity=params.severity,  # type: ignore[arg-type]
            category=params.category,  # type: ignore[arg-type]
            summary=params.summary,
            recommendation=params.recommendation,
            options=[ExceptionOption(**o) for o in (params.options or [
                {"label": "Approve", "action": "approve", "non_revocable": False},
                {"label": "Reject", "action": "reject", "non_revocable": False},
            ])],
            related_policy_refs=[PolicyRef(**r) for r in (params.related_policy_refs or [])],
            bulk_candidate_ids=params.bulk_candidate_ids,
            confidence=params.confidence,
            created_at=time.time(),
        )
        store.upsert_exception(e)
        audit.log("compose-exception.emitted", {"exception_id": e.id, "workflow_id": e.workflow_id})
        return ToolResult(text_result_for_llm=json.dumps({"exception_id": e.id}), result_type="success")

    return compose_exception
